config-service-impl/src/main/java/org/hypertrace/config/service/ConfigServiceGrpcImpl.java

```diff
@@ -0,0 +1,139 @@
+        for (String context : request.getContextsList()) {
+          GetConfigResponse contextSpecificConfigResponse =
+              configStore.getConfig(getConfigResource(request, context), Optional.empty());
+          Value contextSpecificConfig = contextSpecificConfigResponse.getConfig();
+          config = merge(config, contextSpecificConfig);
+        }
+        response = GetConfigResponse.newBuilder()
+            .setConfig(config)
+            .build();
+      }
+
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      log.error("Get config failed for request:{}", request, e);
+      responseObserver.onError(e);
+    }
+  }
+
+  private ConfigResource getConfigResource(UpsertConfigRequest upsertConfigRequest) {
+    return new ConfigResource(upsertConfigRequest.getResourceName(),
+        upsertConfigRequest.getResourceNamespace(),
+        getTenantId(),
+        getActualContext(upsertConfigRequest.getContext()));
+  }
+
+  private ConfigResource getConfigResource(GetConfigRequest getConfigRequest, String context) {
+    return new ConfigResource(getConfigRequest.getResourceName(),
+        getConfigRequest.getResourceNamespace(),
+        getTenantId(),
+        getActualContext(context));
+  }
+
+  private String getActualContext(String rawContext) {
+    return StringUtils.isBlank(rawContext) ? DEFAULT_CONTEXT : rawContext;
+  }
+
+  private String getTenantId() {
+    Optional<String> tenantId = RequestContext.CURRENT.get().getTenantId();
+    if (tenantId.isEmpty()) {
+      throw new IllegalArgumentException("Tenant id is missing in the request.");
+    }
+    return tenantId.get();
+  }
+
+  private Value merge(Value defaultConfig, Value overridingConfig) {
+    if (defaultConfig == null) {
+      return overridingConfig;
+    } else if (overridingConfig == null) {
+      return defaultConfig;
+    }
+
+    // Only if both - defaultConfig and overridingConfig are of kind Struct(Map), then merge
+    // the common fields. Otherwise, just return the overridingConfig
+    if (defaultConfig.getKindCase() == Value.KindCase.STRUCT_VALUE
+        && overridingConfig.getKindCase() == Value.KindCase.STRUCT_VALUE) {
+      Map<String, Value> defaultConfigMap = defaultConfig.getStructValue().getFieldsMap();
+      Map<String, Value> overridingConfigMap = overridingConfig.getStructValue().getFieldsMap();
+
+      Map<String, Value> resultConfigMap = new LinkedHashMap<>(defaultConfigMap);
+      for (Map.Entry<String, Value> entry : overridingConfigMap.entrySet()) {
+        resultConfigMap.put(entry.getKey(),
+            merge(defaultConfigMap.get(entry.getKey()), entry.getValue()));
+      }
+      Struct struct = Struct.newBuilder().putAllFields(resultConfigMap).build();
+      return Value.newBuilder().setStructValue(struct).build();
+    } else {
+      return overridingConfig;
+    }
+  }
+
+  // TODO : get the userId from the context
```

Comment:
I'd recommend omitting this whole concept we have a use case and need. userId itself is ambiguous, as it's unclear what user it represents. Is it the author, the last editor, the user it applies to? All of those are potentially things we might want to track. Also, many configs may driven by the system.

Reply:
`UserId` here represents the user who upserted the config - author/creator and last editor are same as once a config version is inserted into the store, it is never updated/editted. We always create a new version of the config.
I've just kept it here as a placeholder(so that the store can work with it) for now and is not exposed via the API. I think if we need to expose it via the API in future, we can document it to remove ambiguity, right?

Reply:
Fair on edit vs author. I don't believe we use userId anywhere in the platform other than feature flags right now, so we may want to discuss how we want to expose it when we're ready for this. I'd rather not require services to have to start working with the JWT directly.